ethereum/core/src/main/java/org/hyperledger/besu/ethereum/trie/diffbased/common/provider/DiffBasedWorldStateProvider.java

```diff
@@ -120,40 +119,122 @@ public Optional<WorldState> get(final Hash rootHash, final Hash blockHash) {
+   */
   @Override
-  public Optional<MutableWorldState> getMutable(
-      final BlockHeader blockHeader, final boolean shouldPersistState) {
-    if (shouldPersistState) {
-      return getMutable(blockHeader.getStateRoot(), blockHeader.getHash());
+  public Optional<MutableWorldState> getWorldState(final WorldStateQueryParams queryParams) {
+    if (worldStateSharedConfig.isStateful()) {
+      return getStatefulWorldState(queryParams);
     } else {
-      final BlockHeader chainHeadBlockHeader = blockchain.getChainHeadHeader();
-      if (chainHeadBlockHeader.getNumber() - blockHeader.getNumber()
-          >= trieLogManager.getMaxLayersToLoad()) {
-        LOG.warn(
-            "Exceeded the limit of historical blocks that can be loaded ({}). If you need to make older historical queries, configure your `--bonsai-historical-block-limit`.",
-            trieLogManager.getMaxLayersToLoad());
-        return Optional.empty();
-      }
-      return cachedWorldStorageManager
-          .getWorldState(blockHeader.getHash())
-          .or(() -> cachedWorldStorageManager.getNearestWorldState(blockHeader))
-          .or(() -> cachedWorldStorageManager.getHeadWorldState(blockchain::getBlockHeader))
-          .flatMap(worldState -> rollMutableStateToBlockHash(worldState, blockHeader.getHash()))
-          .map(MutableWorldState::freeze);
+      throw new RuntimeException("stateless mode is not yet available");
     }
   }
 
+  /**
+   * Gets the head world state.
+   *
+   * <p>This method returns the head world state, which is the most recent state of the world.
+   *
+   * @return the head world state
+   */
   @Override
-  public synchronized Optional<MutableWorldState> getMutable(
-      final Hash rootHash, final Hash blockHash) {
-    return rollMutableStateToBlockHash(persistedState, blockHash);
+  public MutableWorldState getWorldState() {
+    return headWorldState;
+  }
+
+  /**
+   * Gets the stateful world state based on the provided query parameters.
+   *
+   * <p>This method determines whether to retrieve the stateful world state from the head or from
+   * the cache based on the query parameters. If the query parameters indicate that the world state
+   * should update the head, the method retrieves the stateful world state from the head. Otherwise,
+   * it retrieves the stateful world state from the cache.
+   *
+   * <p>The method follows these steps: 1. Check if the query parameters indicate that the world
+   * state should update the head. 2. If true, call {@link #getStatefulWorldStateFromHead(Hash)}
+   * with the block hash from the query parameters. 3. If false, call {@link
+   * #getStatefulWorldStateFromCache(BlockHeader)} with the block header from the query parameters.
+   *
+   * @param queryParams the query parameters
+   * @return the stateful world state, if available
+   */
+  private Optional<MutableWorldState> getStatefulWorldState(
```

Comment:
I think having Stateful in these method names is redundant.  The worldstate is state, and hence should be stateful.  We are trying to signify whether or not it is backed by a trie or a flat db versus a witness/proof?  

Maybe FullWorldState vs SparseWorldState or CompleteWorldState vs WitnessWorldState ?

Reply:
I think I will do a mix and use FullWorldState vs WitnessWorldState. Seems to be more clear, because SparseWorldState doesn't seems to be a good one